src/core-services/account/mutations/createGroup.js

```diff
@@ -0,0 +1,72 @@
+import _ from "lodash";
+import Logger from "@reactioncommerce/logger";
+import ReactionError from "@reactioncommerce/reaction-error";
+
+/**
+ * @name group/createGroup
+ * @method
+ * @memberof Group/Methods
+ * @summary Creates a new permission group for a shop
+ * It creates permission group for a given shop with passed in roles
+ * @param {object} input - The input supplied from GraphQL
+ * @param {Object} input.groupData - info about group to create
+ * @param {String} input.groupData.name - name of the group to be created
+ * @param {String} input.groupData.description - Optional description of the group to be created
+ * @param {Array} input.groupData.permissions - permissions to assign to the group being created
+ * @param {Array} input.groupData.members - members of the
+ * @param {object} context - The GraphQL context
+ * @param {String} context.shopId - id of the shop the group belongs to
+ * @returns {Object} - `object.status` of 200 on success or Error object on failure
+ */
+export default async function createGroup(input, context) {
+  const { group, shopId } = input;
+  let _id;
+  const {
+    collections: {
+      Groups
+    }
+  } = context;
+
+  // (LEGACY) we are limiting group method actions to only users with admin roles
+  // this also include shop owners, since they have the `admin` role
+  await context.validatePermissionsLegacy(["admin"], null, { shopId });
+
+  // we are limiting group method actions to only users within the account managers role
+  await context.validatePermissions("reaction:account", "create", { shopId });
+
+  const defaultCustomerGroupForShop = Groups.findOne({ slug: "customer", shopId }) || {};
+
+  // TODO: Remove when we move away from legacy permission verification
+  const defaultAdminPermissions = (defaultCustomerGroupForShop.permissions || []).concat("dashboard");
+  const newGroupData = Object.assign({}, group, {
+    slug: context.getSlug(group.name), shopId
+  });
+
+  // TODO: Remove when we move away from legacy permission verification
+  if (!newGroupData.permissions) {
+    newGroupData.permissions = [];
+  }
+
+  // TODO: Remove when we move away from legacy permission verification
+  newGroupData.permissions = _.uniq([...newGroupData.permissions, ...defaultAdminPermissions]);
+
+  // ensure one group type per shop
+  const groupExists = Groups.findOne({ slug: newGroupData.slug, shopId });
+
+  if (groupExists) {
+    throw new ReactionError("conflict", "Group already exist for this shop");
+  }
```

Comment:
I know this is from existing code, but technically it would be best to create a unique index in MongoDB on `slug` + `shopId` fields. That way it's impossible for it to happen. This existing check makes it unlikely but not impossible because two requests for the same slug could come in at virtually the exact same time. If we have the unique index, the `findOne` could be entirely removed and instead throw `Group already exists` error if there's a unique index error from Mongo from the insert below.

Reply:
@aldeed ... with regards to the snippet below
>I know this is from existing code, but technically it would be best to create a unique index in MongoDB on slug + shopId fields

Were you referring to a compound index or two single indexes? ... I have created a compound index but If you clarify it, I will make the necessary change